### PR TITLE
[PF-619] Disable deleting resource when exceeding pool size

### PR DESCRIPTION
**Pull request by yonghaoy**
This is the short-term of 'Primary Job' in Buffer doc:
https://docs.google.com/document/d/1rSIXOrDQYCgptHZIxnxXjM-nHjb_ohxxNGKAl5ZarKw/edit#heading=h.5inn55b6pv26

As Doug suggested "For: down sizing... Currently we don't really need to actively reduce the size"
This also increases the frequency of scheduling service runs.

### CODE REVIEW

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java [hunk 1]
```diff
@@ -11,21 +11,40 @@
...
 
-  /** How often to query for flights to submit. */
-  private Duration flightSubmissionPeriod = Duration.ofSeconds(10);
+  /**
+   * How often to query for flights to submit.
+   *
+   * <p> Sets this to be a larger number because most of the service is deployed in a
+   * multi-instance world, and there will be more than one scheduling service running.
+   */
+  private Duration flightSubmissionPeriod = Duration.ofSeconds(20);
```
Comment (wchamber) [inline src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:38]: why not 3x? Does this need ot scale with the number of instances?

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java [hunk 2]
```diff
@@ -11,21 +11,40 @@
...
+
+  /**
+   * Whether to delete resource when resource count exceeds the pool size.
+   *
+   * <p>This to prevent multiple instances concurrent scheduler deleting extra resources. Also this
+   * is no use case that we need to actively reduce pool size. Resource is only need to be deleted
+   * when pool is inactivated.
+   */
+  private boolean deleteExceedPoolSizeResource;
```
Comment (wchamber) [inline src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:47]: how about `deleteExcessResources`?

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java [hunk 3]
```diff
@@ -11,21 +11,40 @@
...
+   * multi-instance world, and there will be more than one scheduling service running.
+   */
+  private Duration flightSubmissionPeriod = Duration.ofSeconds(20);
+
+  /**
+   * Whether to delete resource when resource count exceeds the pool size.
+   *
+   * <p>This to prevent multiple instances concurrent scheduler deleting extra resources. Also this
+   * is no use case that we need to actively reduce pool size. Resource is only need to be deleted
+   * when pool is inactivated.
```
Comment (wchamber) [inline src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:45]: Should we default this to false? Any thoughts on if we even want to keep the ability to delete excess resources? Would we want to revert this when we have PF-617?

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java [hunk 4]
```diff
@@ -11,21 +11,40 @@
  *
  * <p>Rate limiting: The bottleneck for GCP quota comes from ServiceUsage.batchEnable API. By
  * default, it is 20/100s. Here is how we estimate the time: A Creation flight takes 600~900
- * seconds. If scheduler runs every 10 seconds, in 900 seconds it schedulers runs 70 times. If we
+ * seconds. If a scheduler runs every 10 seconds, in 900 seconds it schedulers runs 70 times. If we
  * have 4 pools, it schedules ({@code resourceCreationPerPoolLimit} * 4) = 4 flights. In total 70
  * times run, 280 flights is scheduled. We estimate this number should works when pool number is
- * 1~10. If we see more errors, we will comeback and revise those configs.
+ * 1~10. If we see more errors, we will comeback and revise those configs. TODO(PF-617): The current
```
Comment (wchamber) [inline src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:17]: nit: add a `<p>` before the TODO

**Review by wchamber — CHANGES_REQUESTED** [review]
I think you should file a ticket for this and discuss why we're doing this short term change and why it's not a good long term solution.

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java [hunk 5]
```diff
@@ -138,7 +159,22 @@ public void scheduleDeactivationFlights_poolDeactivated() throws Exception {
   }
 
   @Test
-  public void scheduleDeactivationFlights_poolSizeReduced() throws Exception {
+  public void scheduleDeactivationFlights_enableDeletion() throws Exception {
+    newPoolWithResourceCount(1, ImmutableMultiset.of(ResourceState.READY, ResourceState.READY));
```
Comment (jaycarlton) [inline src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:163]: why two states if there's only one resource?

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java [hunk 6]
```diff
@@ -138,7 +159,22 @@ public void scheduleDeactivationFlights_poolDeactivated() throws Exception {
...
+
+    PrimaryConfiguration primaryConfiguration = newPrimaryConfiguration();
+    primaryConfiguration.setDeleteExceedPoolSizeResource(true);
+    initializeScheduler(primaryConfiguration);
+
+    Thread.sleep(4000);
```
Comment (jaycarlton) [inline src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:169]: use `TimeUnit.SECONDS(4).sleep()`.

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java [hunk 7]
```diff
@@ -138,7 +159,22 @@ public void scheduleDeactivationFlights_poolDeactivated() throws Exception {
...
+
+    verify(flightManager, times(1))
```
Comment (jaycarlton) [inline src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:171]: you may omit `times(1)` as it's the default.

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java [hunk 8]
```diff
@@ -151,6 +187,18 @@ public void scheduleDeactivationFlights_poolSizeReduced() throws Exception {
     verify(flightManager, never()).submitCreationFlight(any(Pool.class));
   }
 
+  @Test
+  public void notScheduleDeactivationFlights_disableDeletion() throws Exception {
+    newPoolWithResourceCount(1, ImmutableMultiset.of(ResourceState.READY, ResourceState.READY));
+
+    initializeScheduler();
+    Thread.sleep(4000);
```
Comment (jaycarlton) [inline src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:195]: indicate how this time relates to other configured times.

Also, can you mock the clock or thread class, or do you need a real sleep here?

**Review by jaycarlton — APPROVED** [review]
The parts I understand look good. I don't have much context.

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java (hunk 1 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:38]: Changed to be 3x. Feels we don't need to change this frequently every-time number instance change, but we many want to change this if number of instances changes a lot(from 1 to 10?). 
I have a personal env (buffertest) that has multiple(2) instance enabled without this change. Seems it is able to create 1500 resources successfully without little 429 errors.  So I guess 10s still works with 2 instances, changing to 30 is an improvement based on our calculation above.
But all are estimation based on our calculation.

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java (hunk 5 above)
Comment (yonghaoy) [reply src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:163]: pool size is 1, but there are two READY resources, that is why the delete happens

File: src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java (hunk 8 above)
Comment (yonghaoy) [reply src/test/java/bio/terra/buffer/service/resource/FlightSchedulerTest.java:195]: I need a real sleep to make the scheduler really runs.

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java (hunk 3 above)
Comment (wchamber) [reply src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:44]: 2) I don't really follow. If you're trying to figure out what the service does by default, aren't you going to be looking at the code? If we default to false, we don't have to touch any of the `application.yml`, which seems like less understanding of the code.

**Review by wchamber — APPROVED** [review]
lgtm, but please consider changing the default value of `deleteExcessResources` so we don't need to set values in the application yml files.

File: src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java (hunk 3 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/buffer/app/configuration/PrimaryConfiguration.java:44]: This is based on the most recent work with Jay when he moves StairwayCompoment into common-lib.
Having default value for spring config in java code seems not to be a well-know practice. People with spring experience seems will go to appliction.yml to find all values.  Java -> Application/.yaml -> helm is 3 levels lookup. While removing Java having default value, we get one less level. 

But to me, I don't have strong preference, changing to use java file hold default value

**Pull request merged**